Feature/softwareupdate

martonborzak opened the pull request:
Updated software update functionality with new class.
Closes #414
Closes #230 
Closes #422
Closes #419

martonborzak committed:
Initial setup flow update

martonborzak committed:
Softwareupdate class

martonborzak committed:
Update softwareupdate.h

martonborzak committed:
Software update class

martonborzak committed:
Added app version to .pro file and made available in C++ side

martonborzak committed:
Initial settings update

martonborzak committed:
Merge branch 'feature/softwareupdate' of https://github.com/YIO-Remote/remote-software into feature/softwareupdate

martonborzak committed:
Software update UI

martonborzak committed:
Update

martonborzak committed:
Update German translation name

martonborzak committed:
Software update is almost functional now

martonborzak committed:
Renamed update to download in settings

martonborzak committed:
Updated config json schema to fix errors

martonborzak committed:
Update softwareupdate.cpp

martonborzak committed:
Update

martonborzak committed:
Merge branch 'feature/softwareupdate' of https://github.com/YIO-Remote/remote-software into feature/softwareupdate

martonborzak committed:
Software update refinements

martonborzak committed:
Update config.json

zehnm commented on basic_ui/settings/Softwareupdate.qml:
```diff
@@ -21,149 +21,127 @@
+                mouseArea.onClicked: {
+                    if (SoftwareUpdate.updateAvailable) {
+                        if (fileio.exists(appPath + "/downloads/latest.zip")) {
+
```
Isn't here something missing?
Otherwise remove empty block and negate condition.

zehnm commented on qml.qrc:
```diff
@@ -65,5 +64,12 @@
         <file>basic_ui/settings/Wifi.qml</file>
         <file>basic_ui/settings/WifiConnect.qml</file>
+        <file>setup/Setup.qml</file>
+        <file>setup/Step1.qml</file>
```
The names `Step1-5` are very generic. Maybe we'll have another wizard in the future, so it's better to have unique names. E.g. `SetupStep1-5`

zehnm commented on setup/Setup.qml:
```diff
@@ -0,0 +1,47 @@
+
+    SwipeView {
+        id: swipeView
+        width: 480; height: 800
```
Is it possible to use global variables for width & height? This would greatly simplify adapting new displays.

martonborzak replied:
Let's add this to the display hardware and we can take it from there.

zehnm replied:
Ok, the enhanced device configuration is now in develop. I will create a new issue.

martonborzak replied:
Thanks

zehnm commented on setup/Setup.qml:
```diff
@@ -0,0 +1,47 @@
+/******************************************************************************
+ *
+ * Copyright (C) 2018-2019 Marton Borzak <hello@martonborzak.com>
```
It's already 2020 ;-) Applies to other headers too...

zehnm commented on sources/config.cpp:
```diff
@@ -76,6 +75,7 @@ void Config::setConfig(const QVariantMap &config) {
     m_config = config;
     syncConfigToCache();
     emit configChanged();
+    writeConfig();
```
Return value of `writeConfig()` must be propagated to caller. I.e. `Config::setConfig` return type must be changed to `bool`. Otherwise it's fire and forget and an invalid configuration won't be saved without the caller knowing about it!

zehnm replied:
This class has been enhanced in the develop branch.
- Unfortunately setConfig cannot be enhanced with a bool return since it's a Q_PROPERTY
- The client must call isValid() after calling setConfig

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+
+    // start update checker timer
+    m_checkForUpdateTimer = new QTimer(this);
+    m_checkForUpdateTimer->setInterval(3600000);  // 1 hour
```
Make update check interval configurable.
In my opinion 1 hour is quite aggressiv as default value.

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+    qCDebug(CLASS_LC) << "Networkaccessmanager deleted";
+}
+
+void SoftwareUpdate::downloadUpdate(QUrl url) {
```
For a more robust download process, the following points should be considered:
- If download file already exists it's better to delete it first.
- Check free space, only download if there's enough space available (e.g. 100 MB).
- Download into a temp file and rename after successful download.
  This eliminates race conditions if another process polls the download directory.
- Download archive should be verified with hash. This requires publishing the zip hash on the server.

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+            combinedNewVersion = (none.append(ntwo).append(nthree)).toInt();
+
+            QStringList currentVersionDigits = m_currentVersion.split(".");
+            QString     cone                 = QString("%1").arg(currentVersionDigits[0].toInt(), 2, 10, QChar('0'));
```
Correct names are: major, minor, patch

zehnm commented on sources/softwareupdate.h:
```diff
@@ -0,0 +1,117 @@
+    qint64  m_bytesTotal    = 0;
+    QString m_downloadSpeed;
+
+    QString                m_updateUrl = "https://update.yio.app/update";
```
Update url should be configurable

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+                Notifications::getInstance()->add(
+                    false, tr("New software is available"), tr("Download"),
+                    [](QObject *param) {
+                        SoftwareUpdate *i = qobject_cast<SoftwareUpdate *>(param);
```
Variable name `i` is commonly used as counter variable. Please use descriptive variable names whenever possible.

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+    emit downloadComplete();
+}
+
+void SoftwareUpdate::startUpdate() {
```
As I understand the implemented functionality, this method only downloads the latest version and doesn't start the installation. Correct?
In this case the method name is misleading and should be renamed to `startDownload` or `initiateDownload`.
Where is the actual installation of the update? Afaik there's a shell script somewhere and in the screen it says the update is installed between 3am-5am. But where is that implemented?

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+    request.setRawHeader("Accept", "application/json");
+
+    QJsonObject json;
+    json.insert("version", m_currentVersion);
```
We have to add a device type parameter to the update request:
- Update server needs to return the correct download for the client runtime. We don't want to install an ARM runtime on a Windows Desktop!
- Furthermore we'd probably like to distinguish a real YIO remote from all the test and desktop installations.

Possible implementation:
- `?type=remote` if running on the YIO remote
- `?type=macos|win|linux` elsewhere. 

Or something similar...

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+            Notifications::getInstance()->add(true, tr("Software update: ") + jsonObject["error"].toString());
+        } else {
+            m_downloadUrl = jsonObject["url"].toString();
+            m_newVersion  = jsonObject["latest"].toString();
```
Do we only publish releases or do we also allow a 'development channel', i.e. daily builds / snapshots?
Then we need to append a timestamp or build number to distinguish snapshot versions.
Might be could to already implement the optional suffix. But probably better to discuss this first...

zehnm commented on sources/softwareupdate.cpp:
```diff
@@ -0,0 +1,239 @@
+    m_download = m_manager->get(request);
+
+    connect(m_download, &QNetworkReply::downloadProgress, this, &SoftwareUpdate::downloadProgress);
+    connect(m_download, &QNetworkReply::finished, this, &SoftwareUpdate::downloadFinished);
```
Error handling is missing if download fails for any reasons!
The QNetworkReply::error(QNetworkReply::NetworkError code) signal must handled as well.

zehnm commented on config-schema.json:
```diff
@@ -47,19 +49,21 @@
+                  "Hallway light"
                 ],
-                "pattern": "^(.*)$"
+                "pattern": ""
```
Empty pattern makes no sense. Must be removed, otherwise the validation might fail depending on the used library!
Applies to all pattern definitions in the schema.

zehnm commented on config-schema.json:
```diff
@@ -341,18 +352,21 @@
                         "0000,006C,0000,0022,00AD,00AD,0016,0041,0016,0041,0016,0041,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0041,0016,0041,0016,0041,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0041,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0016,0041,0016,0016,0016,0041,0016,0041,0016,0041,0016,0041,0016,0041,0016,0041,0016,06FB"
                       ],
-                      "pattern": "^(.*)$"
+                      "pattern": ""
```
Here a pattern makes sense and should be defined.
An IR code must be in a defined format, otherwise it cannot be processed.

zehnm commented on config-schema.json:
```diff
@@ -708,15 +735,21 @@
+              ],
               "title": "Log level",
-              "default": "INFO"
+              "default": ""
```
A default value is required for the log level!
Set to "INFO"

zehnm commented on config-schema.json:
```diff
@@ -708,15 +735,21 @@
               "type": "boolean",
               "title": "Log to console",
-              "default": true
+              "default": false
```
This disables the journald log. Is this really what we want?

zehnm commented on config-schema.json:
```diff
@@ -727,16 +760,16 @@
-              "minimum": 0,
               "title": "Purge log files in hours",
-              "default": 12
+              "default": 0,
```
Attention: does this disable the log file purging or does it immediately delete the log files whenever it checks if files need to be deleted?
Behaviour should be documented :-)

zehnm commented on config-schema.json:
```diff
@@ -727,16 +760,16 @@
-              "minimum": 0,
               "title": "Log message queue size for websocket API",
-              "default": 100
+              "default": 0,
```
Does this disable the log queue? I.e. no logs over the websocket API?

zehnm commented on config-schema.json:
```diff
@@ -752,27 +785,28 @@
           "type": "integer",
-          "title": "Shutdown time in minutes",
-          "default": 21600
+          "title": "Shutdown time in seconds",
```
Why seconds?
I personally find minutes more convenient in a configuration file. No need to specify seconds.

martonborzak replied:
This is never shown like this to the user. `config.json` is only edited by us for now.

zehnm replied:
The config schema should be used by the web configurator to initialize the settings. Therefore a sensible default value is required. Otherwise it has no use in the schema file and must be removed.

zehnm commented on config-schema.json:
```diff
@@ -752,27 +785,28 @@
-          "title": "WiFi switch off time in minutes",
-          "default": 3600
+          "title": "WiFi switch off time in seconds",
+          "default": 0
```
Shouldn't we switch off wifi by default? Of course we need to keep it on during the initial setup.
Why seconds?
I personally find minutes more convenient in a configuration file. No need to specify seconds.

martonborzak replied:
This is never shown like this to the user. `config.json` is only edited by us for now.

zehnm reviewed (commented):
Most important things to improve is the file download handling:
- configurable download url
- delete old file before download
- free space check before download
- using temp file and rename if successfully downloaded
- handling QNetworkReply::error(QNetworkReply::NetworkError code)

The config json schema file needs some refining too. Probably better to create a dedicated task for that.

zehnm committed:
WifiControl signals for connection and join status

Added new signals:
- connected();
- disconnected();
- joinError(WifiControl::JoinError error);

The join() operation is now asynchronous and uses the new signals to communicate
with the caller.

zehnm commented:
The script `scripts/softwareupdate.js` should be deleted and removed from qml.qrc.

martonborzak committed:
Update

martonborzak committed:
Update

martonborzak committed:
Removed unused javascript

martonborzak committed:
Update StatusBar.qml

martonborzak committed:
Update integrations.cpp

zehnm commented on setup/SetupStep5.qml:
```diff
@@ -0,0 +1,69 @@
+        }
+
+        onJoinError: {
+            console.debug("Wifi error:" + error);
```
Shouldn't the user be informed that the network couldn't be joined?
Or is there another feedback, e.g. a timeout handling?

zehnm committed:
Merge branch 'develop' into feature/softwareupdate

zehnm committed:
Fixed double include

martonborzak committed:
Setup flow for testing

martonborzak committed:
UI tweaks

martonborzak committed:
api update for hostname generation

martonborzak committed:
Create locale.json

martonborzak committed:
Added country selector to setup flow

martonborzak committed:
Keyboard fix

martonborzak committed:
Added multiple keyboards for languages

martonborzak committed:
Removed unused file

martonborzak committed:
Typo fix

martonborzak committed:
Updated licenses

martonborzak committed:
Typo fix

martonborzak committed:
Get MAC address with QNetworkInterface

martonborzak committed:
Update yioapi.cpp

martonborzak committed:
API class cleanup

martonborzak committed:
Avoid creating duplicate notifications

martonborzak committed:
Updated Style singleton qml

zehnm committed:
Initial software update enhancements [wip]

- Added configuration object in config.json and passing it to SoftwareUpdate
- Added storage check
- Added simple device identification
- Downloading into temp file
- Refactored constructor
- Added TODOs for remaining tasks

martonborzak committed:
Removed "v" prefix from version number on server side

martonborzak committed:
Added `installAvailableChanged` signal

zehnm committed:
Software update refactoring

- moved download into dedicated class
- improved error handling
- more TODOs...

zehnm committed:
Fixed cpplint issues

zehnm committed:
Software update enhancements

- Queueable file downloads
- Metafile for version information
- Various fixes

zehnm committed:
Software update enhancements

- Using latest update API with GET request
- Limit archives to .tar and .gz
- Correctly handling response status codes

zehnm committed:
fixup! Software update enhancements

martonborzak committed:
Update notifications.cpp

martonborzak committed:
Merge branch 'feature/softwareupdate' of https://github.com/YIO-Remote/remote-software into feature/softwareupdate

zehnm committed:
Using YIO_PLUGIN_DIR env var for plugin path

zehnm committed:
Merge branch 'develop' into feature/softwareupdate

martonborzak committed:
Update notifications.cpp

martonborzak committed:
Merge branch 'feature/softwareupdate' of https://github.com/YIO-Remote/remote-software into feature/softwareupdate

zehnm committed:
Fixed potential program crash

The underlying collection of an iterator may never be modified during iteration

zehnm committed:
cpplint line length issue

zehnm committed:
Separated device detection from SoftwareUpdate

zehnm committed:
Improved environment detection, initial app update

zehnm committed:
Fixed RPi device detection

zehnm committed:
Retrieve app version from git tag

This closes #406

zehnm committed:
Merge branch 'feature/softwareupdate' of https://github.com/YIO-Remote/remote-software into feature/softwareupdate

zehnm committed:
Logging plugin version from metadata

zehnm committed:
Enhanced crosscompile build

- Create md5 sums of all archive files
- Fetch git tags and history to determine version number
- New artifact naming
- Added build timestamp in archive
- Status batches in README

zehnm committed:
fixup! Enhanced crosscompile build

zehnm committed:
Simplified app update check

Aligned to new yio-updateserver API

martonborzak committed:
Fix for #422

martonborzak committed:
Caching images

zehnm committed:
Added Notification operator <<

zehnm committed:
Software update enhancements

- Changes for latest update server API
- Properly encoding device name query parameter
- Don't send unknown remote-os version
- Removed file system readonly check

martonborzak committed:
Update notifications.cpp

martonborzak committed:
Battery graph x axis fix

martonborzak committed:
Merge branch 'feature/softwareupdate' of https://github.com/YIO-Remote/remote-software into feature/softwareupdate

martonborzak committed:
YIO API hostname fix

martonborzak committed:
Icons update

martonborzak committed:
Fix for #419

martonborzak committed:
Possible fix for #420

Possible solution where we hide the item

martonborzak committed:
Possible fix for #420

martonborzak committed:
Update Wifi.qml

martonborzak committed:
Cleanup
